dotnetLearning Update

HoroshoNaVolge opened the pull request:
Lesson 2 finished.

HoroshoNaVolge committed:
structure update

HoroshoNaVolge committed:
internal commit

HoroshoNaVolge committed:
some shitty changes

HoroshoNaVolge committed:
remake of FactoryApp

Structure remade.
DI and configuration used.
Tried MVC

HoroshoNaVolge committed:
Update Program.cs

navferty commented on src/dotnetLearning/FactoryApp/Service/FacilityService/JsonContainer.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    internal class JsonContainer(IEnumerable<Factory> factory, IEnumerable<Unit> unit, IEnumerable<Tank> tank)
+    {
+        public IEnumerable<Factory> Factory { get; set; } = factory;
```
Дело вкуса наверно, но я в моделях обычно предпочитаю коллекции объявлять как `IReadOnlyCollection` или `IList`, за редкими исключениями: сценарий, когда коллекция ещё не материализована, довольно редки. Например, даже если читаем потоком массив из json массива из файла, всё равно `IEnumerable`/`IAsyncEnumerable` получаем только на верхнем уровне, а сами элементы, например, будут уже десериализованы каждый полностю.
Зато плюсом получаем возможность логирования количества элементов в модели (`_logger.LogTrace("Загружено {Count} заводов.", item.Factories.Count)`)

navferty replied:
Также я бы предложил использовать множественное число для свойства-коллекции: `Factories`, `Units` etc.

navferty commented on src/dotnetLearning/FactoryApp/Service/FactoryAppService.cs:
```diff
@@ -0,0 +1,66 @@
+            if (options.Value.FacilitiesJsonFilePath is not null)
+                await facilityService.DeserializeDataJson(options.Value.FacilitiesJsonFilePath);
+            else
+                throw new ArgumentNullException("Ошибка в файле конфигурации appsettings.json");
```
Напомню, метод `string.IsNullOrEmpty` позволит сразу проверить не только на null, но и на пустую строку. Опять же вкусовщина, но можно немного упростить вложенность:

```suggestion
            if (string.IsNullOrEmpty(options.Value.FacilitiesJsonFilePath))
                throw new ArgumentNullException("Ошибка в файле конфигурации appsettings.json");

            await facilityService.DeserializeDataJson(options.Value.FacilitiesJsonFilePath);
```

navferty commented on src/dotnetLearning/FactoryApp/View/IView.cs:
```diff
@@ -0,0 +1,7 @@
+namespace dotnetLearning.FactoryApp.View
+{
+    public interface IView{
```
Мелочи, но я бы рекомендовал следить за единообразием оформления кода в проекте: фигурные скобки, отступы, стиль именования (PascalCase/camelCase), наличие/отсутствие пустой строки в конце файла и т.д.

navferty commented on src/dotnetLearning/Program.cs:
```diff
@@ -1,25 +1,34 @@
-        //foreach (var i in new IEnumerableLearn()) { await Console.Out.WriteLineAsync(i.ToString()); }
-        //foreach (var i in new IEnumerableLearnWithOwnEnumerator()) { await Console.Out.WriteLineAsync(i.ToString()); }
-    }
+         IConfigurationRoot config = new ConfigurationBuilder()
```
Кстати, довольно любопытный проект для разработки консольных приложений схожим способом как и asp.net: [Cocona](https://github.com/mayuki/Cocona). [Обзор от Nick Chapsas](https://www.youtube.com/watch?v=4KHFrbJHCHc&ab_channel=NickChapsas).
Не то чтобы предлагаю тут же всё переписать, но можно поэкспериментировать: возможно, на будущее пригодится =)

HoroshoNaVolge committed:
Lesson0 updated

navferty commented on src/dotnetLearning/FactoryApp/Model/Factory.cs:
```diff
@@ -0,0 +1,23 @@
+            Name = name;
+            Description = description;
+        }
+        public int? Id { get; set; }
```
Имеет ли смысл сущность `Factory`, у которой не определён `Id`? Возможно, имеет смысл отказаться от nullability свойств `Id` и `Name`. В C# есть возможность пометить свойство required, чтобы при создании объекта инициализация этих свойств была обязательна. Кроме того, можно изменить сеттер на `init`, чтобы запретить повторное присваивание значения этому свойству:

```c#
public required int Id { get; init; }
public required string Name { get; init; }
```

В таком случае, наличие дополнительного контруктора с аргументами становится уже излишним: все required-свойства должны быть проинициализированы сразу при создании объекта:

```c#
var factory = new Factory
{
  Id = 42,
  Name = "Sample", // если не проинициализировать required-свойство, код не компилируется
}
```

navferty commented on src/dotnetLearning/FactoryApp/Service/FacilityService/FacilityService.cs:
```diff
@@ -0,0 +1,117 @@
+        private JsonContainer? _container = null;
+        private static readonly JsonSerializerOptions _jsonSerializerOptions = new()
+        {
+            Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
```
В чём необходимость использования кодировщика не по умолчанию?

Если требуется избежать кодирования только какого-то подмножества символов юникода, можно явно задать character set'ы, которые не будут экранироватся, как в [этом примере](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/character-encoding#serialize-language-character-sets):

```c#
var options = new JsonSerializerOptions
{
    Encoder = JavaScriptEncoder.Create(UnicodeRanges.BasicLatin, UnicodeRanges.Cyrillic),
    WriteIndented = true,
};
```

navferty commented on src/dotnetLearning/FactoryApp/Service/FacilityService/FacilityService.cs:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        public Unit? FindUnit(string tankName) =>
+            _units?.FirstOrDefault(u => _tanks?.Any(t => t.Name == tankName && t.UnitId == u.Id) == true);
```
`== true` можно опустить =)

HoroshoNaVolge replied:
тогда нужено явно проверять _tanks на null, до метода. Если правильно понимаю... или менять на ?? false

HoroshoNaVolge committed:
add excel export and comments resolved

HoroshoNaVolge committed:
excel import and export update

plus removed camelCase

HoroshoNaVolge committed:
userInteractionAndSearchMethod Added

HoroshoNaVolge committed:
userInput Event added

HoroshoNaVolge committed:
Update ConsoleView.cs

HoroshoNaVolge committed:
search update LINQ both syntaxes

method chain and qurey exp

navferty commented on src/dotnetLearning/FactoryApp/Service/ExcelTransformator.cs:
```diff
@@ -0,0 +1,130 @@
+                {
+                    package.SaveAs(new FileInfo(excelFilePath));
+                }
+                catch (InvalidOperationException)
```
если предполагается, что этот перехват исключения поможет обработать ситуацию, когда файл с таким именем уже существует - лучше предварительно явно проверить наличие файла в папке (например через `if (File.Exists(filename))`, вместо ловли исключения.

navferty commented on src/dotnetLearning/FactoryApp/Service/ExcelTransformator.cs:
```diff
@@ -0,0 +1,130 @@
+            foreach (var item in items)
+            {
+#nullable disable
+                var properties = item.GetType().GetProperties();
```
Использование рефлексии может быть опасным: например, если в ходе рефакторинга изменится порядок свойств, или со временем добавятся новые свойства - то формат файла может незаметно и неожиданно для разработчика измениться.

В данной ситуации, я бы предложил сделать для каждой сущности, записываемой в файл, отдельный метод/класс, который будет отвечать за получение набора значений (например в виде `Dictionary<string, object?>`), а также набора ключей. Эти методы можно объявить прямо в классах-моделях, либо в отдельных вспомогательных классах, например так:

```c#
static class EntityExcelSerializationHelper
{
  static IDictionary<string, object?> GetValues(this Factory factory) => /* get values by each key for provided object */;
  static IList<string> GetFactoryKeys() => /* get all keys for Factory */

  /* other entity types' helpers */
}
```

navferty commented on src/dotnetLearning/FactoryApp/Service/ExcelTransformator.cs:
```diff
@@ -0,0 +1,130 @@
+                    }
+
+                    // потому что читает число как double
+                    catch (ArgumentException)
```
можно избежать этого исключения, если сразу проверить тип текущего свойства

navferty commented on src/dotnetLearning/FactoryApp/Service/FactoryAppService.cs:
```diff
@@ -1,33 +1,147 @@
-            //view.ShowMessage(facilityService.GetUnitsSummary());
+            var userInteractionFinished = false;
+#nullable disable
+            view.InputReceived += HandleUserInput;
```
При оформлении подписки на событие следует помнить, что от него [необходимо отписываться](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/events/how-to-subscribe-to-and-unsubscribe-from-events#unsubscribing). В частности, отсуствие отписки может привести к тому, что более не нужные объекты могут оставаться достижимыми, и не быть собранными сборщиком мусора.

В данном случае, поскольку оба сервиса (и подписчик, и продьюсер событий) зарегистрированы как синглтоны, это не должно вызвать проблем, но на будущее рекомендую почитать [эту статью](https://www.spicelogic.com/Blog/net-event-handler-memory-leak-16), про ситуацию где подписка на событие без отписки от него может привести к проблемам.

navferty commented on src/dotnetLearning/FactoryApp/View/ConsoleView.cs:
```diff
@@ -2,7 +2,28 @@ namespace dotnetLearning.FactoryApp.View
+            var userInputEventArgs = new UserInputEventArgs
+            {
+                UserInput = userInput,
+                InputTime = DateTime.Now
```
Я бы предложил использовать `DateTime.UtcNow`, либо `DateTimeOffset` (если важно сохранить информацию о часовом поясе). Например, если в будущем эти события, например, будут сохраняться в БД (особенно на удалённом сервере), может быть непонятно, в каком поясе было событие ввода у разных пользователей.

HoroshoNaVolge committed:
private commit

HoroshoNaVolge committed:
excelvwriter update

HoroshoNaVolge committed:
CRUD serialization added

HoroshoNaVolge committed:
minor changes

Todo: get rid of duplicate code
CRUD logic finished.
Next step: Entity Framework

HoroshoNaVolge committed:
EntityFramework DB add initial commit

First step:
Registration of service done
Create (CRUD) - export all objects to postgres done.

HoroshoNaVolge committed:
Db CRUD logic update

При операциях добавления/ удаления в БД или Json или Excel, сами данные в модели программы должны удаляться/добавляться? Или это должно быть изолировано от самой программы.

HoroshoNaVolge committed:
CRUD excel and json update

HoroshoNaVolge committed:
ISerializationService update

HoroshoNaVolge committed:
update CRUD and structure
